pkg/controller/kubeconfig.go

```diff
@@ -6,60 +6,111 @@ import (
 
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/util/rand"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
 const (
 	secretTypeBootstrapToken v1.SecretType = "bootstrap.kubernetes.io/token"
+	machineNameLabelKey                    = "machine.k8s.io/machine.name"
+	tokenIDKey                             = "token-id"
+	tokenSecretKey                         = "token-secret"
+	tokenFormatter                         = "%s.s"
 )
 
+func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
+	token, err := c.createBootstrapToken(name)
+	if err != nil {
+		return nil, err
+	}
+
+	infoKubeconfig, err := c.kubeconfigProvider.GetKubeconfig()
+	if err != nil {
+		return nil, err
+	}
+
+	outConfig := infoKubeconfig.DeepCopy()
+
+	outConfig.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		"": {
+			Token: token,
+		},
+	}
+
+	return outConfig, nil
+}
+
 func (c *Controller) createBootstrapToken(name string) (string, error) {
+	existingSecret, err := c.getSecretIfExists(name)
+	if err != nil {
+		return "", err
+	}
+	if existingSecret != nil {
+		return c.updateSecretExpirationAndGetToken(existingSecret)
+	}
+
 	tokenID := rand.String(6)
 	tokenSecret := rand.String(16)
 
 	secret := v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Name:   fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Labels: map[string]string{machineNameLabelKey: name},
```

Comment:
this will work only for new `Secrets`. In the worst case scenario we will create additional `Secret` next to existing.